### ISSUE TITLE
AWS.SimpleQueueService.BatchEntryIdsNotDistinct

### ISSUE THREAD

**Issue by benkeil** [dashbitco/broadway_sqs#63]
When using non FIFO queues, it happens quite often that I receive the same message more than once. That's ok during processing, but when broadway wants to acknowledge the messages (which are in the same batch) if throws an error and move the messages to the DLQ.

Could we just filter out duplicates during acknowledging?

**Comment by josevalim** [comment]
Hi @benkeil! I have moved this to the SQS repo. I think we can support a new option ack type, called `:ack_duplicate`. here are the docs: https://github.com/dashbitco/broadway_sqs/blob/master/lib/broadway_sqs/producer.ex#L28

Would you link to send a PR for that?

**Comment by benkeil** [comment]
I will do

**Comment by benkeil** [comment]
Not sure how to handle the configuration part. If we add the new type `ack_duplicate` you can set e.g. `on_success: :ack` and `on_failure: :ack_duplicate` which makes no sense. Wouldn't it make more sense to leave the options with `:ack` and `:noop` and add a new option `:ack_mode` with `:default` and `:duplicate`?

**Comment by josevalim** [comment]
@benkeil why do you say it makes no sense? Maybe it has no practical use but in theory there is no problem in allowing success to not ack duplicates and allow failures to ack so.

**Comment by josevalim** [comment]
To give such an example, consider this:

1. You want to always mark messages as acked
2. Your pipeline checks for the SQS message ID and it raises if the IDs are duplicate

In this case, you want to have exactly `on_success: :ack, on_failure: :ack_duplicate`, because you know dupes will fail (and therefore should ack) and you expect successes to never have dupes.

**Comment by benkeil** [comment]
Don't get your point. Either you want to ack or you don't. And if you want to ack you can't ack the same message I'd twice in the same batch, do you Must skip it to not get errors. Also from a Business Logic side it makes no sense, because it's the same message.

**Comment by josevalim** [comment]
I see your point. `:ack_duplicate` is a poor name. We can change the name but I still think it should be an option that goes alongside `:ack | :noop` because that's what provides most flexibility.

It is also important to notice that we can't see it is a duplicate until we ack, so it is more like `:ack_and_ignore_duplicate_errors`.

**Comment by benkeil** [comment]
You see at the very beginning if it is a duplicate on the message id. Theoretically you can skip the processing at all.
What is the purpose of being so flexible? It is not possible to ack these messages in the same batch. You could ack the failed messages separate from the succes messages, but what if you receive the message 3 times?

**Comment by josevalim** [comment]
Oh, I see. If they are duplicate in the same batch, can it be the producer could have filtered them too? In any case, I think having a separate option called “filter_duplicate_ids_in_batch” will suffice. Should it be false or true by default?

**Comment by benkeil** [comment]
I think if you misconfigured the queue with a too low visibility_timeout, this could also happen.
So I would suggest activate it by default but log an info/warning and/or add the number of duplicates to the metrics.
What exactly happens if you filter them out on SQS side?

**Comment by josevalim** [comment]
>  So I would suggest activate it by default but log an info/warning and/or add the number of duplicates to the metrics.

Let's go this approach (logging), we don't even need an option. Thanks for all the discussion so far!

**Comment by patrickberkeley** [comment]
For what it's worth, we too are running into this issue.

**Comment by HeavyBR** [comment]
Hello guys,

I was trying to understand this problem (this issue and #49), just to see if I have understand the discussion, the idea here is to filter out the duplicate messages during the acknowledgment and log a warning on the filtered messages by default, it is it?

**Comment by josevalim** [comment]
Correct! Perhaps I would make it opt-in though, so people decide if changing their settings may be better.

**Comment by HeavyBR** [comment]
One thing that I'm thinking about is that as mentioned in issue #49, we have this advise in the SQS docs talking about deleting duplicates:

> If you receive a message more than once, each time you receive it, you get a different receipt handle. You must provide the most recently received receipt handle when you request to delete the message (otherwise, the message might not be deleted).
> https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-queue-message-identifiers.html

Should we be concerned about that? Because if yes, I didn't found a way to get the receive timestamp of the messages, any tips?

**Comment by josevalim** [comment]
@HeavyBR perhaps it needs to be explicitly requested in the `:metadata` field?

**Comment by ndrluis** [comment]
I was experiencing the AWS.SimpleQueueService.BatchEntryIdsNotDistinct error multiple times and was wondering why it was happening so frequently. After investigating the issue, I realized that the problem could be due to the visibility timeout configuration being insufficient for my use case. In my particular scenario, my consumer takes around an hour to process messages, and I was having this error frequently.

After checking the [AWS documentation](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html), I found that the recommended solution is to increase the visibility timeout to a value greater than the time it takes for the consumer to process the message. So, I increased the visibility timeout to 90 minutes, and the error disappeared.

**Comment by maxsalven** [comment]
I've run into this issue as well, and in our case I don't believe it to be visibility timeout related.

In case it helps anyone, Claude created this following solution:
```elixir
defmodule Push.Broadway.SQSDedupAck do
  @moduledoc """
  Acknowledger wrapper around `BroadwaySQS.ExAwsClient` that dedupes
  messages by SQS `message_id` before delegating to the underlying
  `DeleteMessageBatch`.

  Standard SQS queues provide at-least-once delivery: the same physical
  message can be received more than once, each time with a different
  `receipt_handle` but the same `message_id`. `BroadwaySQS.ExAwsClient`
  uses `message_id` as the batch entry `Id` in `DeleteMessageBatch`,
  so two deliveries of the same message landing in one ack batch get
  rejected wholesale with `AWS.SimpleQueueService.BatchEntryIdsNotDistinct`.

  See https://github.com/dashbitco/broadway_sqs/issues/63.

  Dedup rules:
    * If the same `message_id` appears multiple times, keep the *last*
      occurrence (latest receipt). Older receipts may be invalid if
      visibility expired between deliveries.
    * If a `message_id` appears in both `successful` and `failed`, favor
      `successful` — we did process the event at least once, so delete it.
  """

  @behaviour Broadway.Acknowledger

  @impl true
  def ack(ack_ref, successful, failed) do
    {successful, failed} = dedupe(successful, failed)
    BroadwaySQS.ExAwsClient.ack(ack_ref, successful, failed)
  end

  @doc """
  Applies the dedupe rules to a pair of `{successful, failed}` message lists.
  Exposed for testing; `ack/3` wraps this and delegates to the real acker.
  """
  def dedupe(successful, failed) do
    successful = dedupe_keep_last(successful)
    successful_ids = MapSet.new(successful, &message_id/1)

    failed =
      failed
      |> Enum.reject(&(message_id(&1) in successful_ids))
      |> dedupe_keep_last()

    {successful, failed}
  end

  @impl true
  def configure(ack_ref, ack_data, options) do
    BroadwaySQS.ExAwsClient.configure(ack_ref, ack_data, options)
  end

  @doc """
  Swaps the acknowledger module on a `BroadwaySQS.ExAwsClient`-acked
  message so that `ack/3` routes through this dedupe wrapper.
  """
  def wrap(%Broadway.Message{acknowledger: {BroadwaySQS.ExAwsClient, ack_ref, ack_data}} = msg) do
    %{msg | acknowledger: {__MODULE__, ack_ref, ack_data}}
  end

  def wrap(%Broadway.Message{} = msg), do: msg

  defp dedupe_keep_last(messages) do
    messages
    |> Enum.reverse()
    |> Enum.uniq_by(&message_id/1)
    |> Enum.reverse()
  end

  defp message_id(%Broadway.Message{acknowledger: {_, _, %{receipt: %{id: id}}}}), do: id
end
```

And then in your `handle_message/3`:
```elixir
  def handle_message(:default, %Message{} = message, _context) do
    message
    |> Push.Broadway.SQSDedupAck.wrap()
    |> Push.Broadway.decode_data()
    |> set_batch_key()
  end
```

It would be great if the fix was built into the library. FWIW, Claude's thoughts on that:

### The root flaw in broadway_sqs
In deps/broadway_sqs/lib/broadway_sqs/ex_aws_client.ex:80-83:

```elixir
defp build_acknowledger(message, ack_ref) do
  receipt = %{id: message.message_id, receipt_handle: message.receipt_handle}
  {__MODULE__, ack_ref, %{receipt: receipt}}
end
```

They chose to use the SQS `message_id` as the batch entry `Id`. But per the [DeleteMessageBatch API](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_DeleteMessageBatch.html), the Id is caller-chosen and only needs to be unique within the batch — it's for correlating success/failure responses, not for identifying the message. SQS identifies the message by `ReceiptHandle`.

#### The ideal upstream fix
Change `delete_messages/2` to use local indices:

```elixir
defp delete_messages(messages, ack_options) do
  receipts =
    messages
    |> Enum.with_index()
    |> Enum.map(fn {msg, idx} ->
      {_, _, %{receipt: %{receipt_handle: rh}}} = msg.acknowledger
      [id: "#{idx}", receipt_handle: rh]
    end)

  ack_options.queue_url
  |> ExAws.SQS.delete_message_batch(receipts)
  |> ExAws.request!(ack_options.config)
end
```

That's the whole fix. `build_acknowledger` can drop the `:id` field entirely (or keep it as metadata if anything else reads it).
